enf/resource_enf_domain_ratelimit.go

```diff
@@ -0,0 +1,176 @@
+package enf
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/xaptum/go-enf/enf"
+)
+
+func resourceEnfDomainRateLimit() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceEnfDomainRateLimitCreate,
+		Read:   resourceEnfDomainRateLimitRead,
+		Update: resourceEnfDomainRateLimitUpdate,
+		Delete: resourceEnfDomainRateLimitDelete,
+
+		Schema: map[string]*schema.Schema{
+			"domain": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"packets_per_second": {
+				Type:     schema.TypeInt,
+				Required: true,
+			},
+			"packets_burst_size": {
+				Type:     schema.TypeInt,
+				Required: true,
+			},
+			"bytes_per_second": {
+				Type:     schema.TypeInt,
+				Required: true,
+			},
+			"bytes_burst_size": {
+				Type:     schema.TypeInt,
+				Required: true,
+			},
+			"type": {
+				Type:     schema.TypeString,
+				Optional: true,
+				Default:  "default",
+			},
+		},
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+	}
+}
+
+func resourceEnfDomainRateLimitCreate(d *schema.ResourceData, meta interface{}) error {
+	// The ENF automatically creates a rate limit for each domain, so just update the values.
+	return resourceEnfDomainRateLimitUpdate(d, meta)
+}
+
+func resourceEnfDomainRateLimitRead(d *schema.ResourceData, meta interface{}) error {
+	switch d.Get("type") {
+	case "default":
+		return domainDefaultRateLimitRead(d, meta)
+	case "max":
+		return domainMaxRateLimitRead(d, meta)
+	default:
+		return fmt.Errorf("Invalid rate limit type %s. The type must be either default or max", d.Get("type"))
+	}
+}
+
+func domainDefaultRateLimitRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*EnfClient).Client
+	domainRateLimit, resp, err := client.Domains.GetDefaultEndpointRateLimits(context.Background(), d.Get("domain").(string))
+	if err != nil {
+		if resp.StatusCode == http.StatusNotFound {
+			d.SetId("")
+			return nil
+		}
+		return err
+	}
+
+	d.SetId(d.Get("domain").(string))
+	d.Set("packets_per_second", domainRateLimit.PacketsPerSecond)
+	d.Set("packets_burst_size", domainRateLimit.PacketsBurstSize)
+	d.Set("bytes_per_second", domainRateLimit.BytesPerSecond)
+	d.Set("bytes_burst_size", domainRateLimit.BytesBurstSize)
+
+	return nil
+}
+
+func domainMaxRateLimitRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*EnfClient).Client
+	domainRateLimit, resp, err := client.Domains.GetMaxDefaultEndpointRateLimits(context.Background(), d.Get("domain").(string))
+	if err != nil {
+		if resp.StatusCode == http.StatusNotFound {
```

Comment:
Do we need this check?  The rate limits should always be found.  If not, it's a more serious and won't be a 404.

Reply:
I see your point - we might as well just return the error in the case when it's a 404. As a follow-up, can I ask why you chose to check for a 404 [here](https://github.com/xaptum/terraform-provider-enf/blob/master/enf/resource_enf_firewall_rule.go#L119) for the firewall resource Read?

Reply:
Firewall rules can be deleted.  So that code checks if the user deleted the rule outside of Terraform (say using the enfcli) and if so, informs Terraform that the rule was deleted.